Cannot Prune Users / Features

### What version of Node.js are you using?

Latest

### What operating system are you using?

Windows Server 2022

### What version of SnailyCAD are you currently using?

v1.79.1 66e1af8

### Describe the Bug

Pruning will not work. We've got 4100 users and every time I try to prune it simply won't prune any. I will usually try 6 months inactive range. Can we manually trim users in the DB? Is this some issue with foreign key checks or what? It just gives a network error. I'm assuming maybe because there are so many that its trying to prune at once that it times out or something? Its really irritating.

![image](https://github.com/SnailyCAD/snaily-cadv4/assets/25211271/5ae36da0-284e-4ae0-a813-ce88363fbfc9)

Additionally, I wanted to suggest ability to Citizens maybe if they haven't been used in months?
### Expected Behavior

To be able to prune users.

### To Reproduce

_No response_

I am able to reproduce this @WhitigolProd on v1.79.1 47faaa4
  


```

Prismaclientknownrequesterror ( Invalid `Prisma.User.Delete()` Invocation: Foreign Key Constraint Failed On The Field: `PetMedicalRecord_petId_fkey (Index)`)

An unexpected error occurred
```

SnailyCAD Console Log:

`[1]
[1] SnailyCADv4 is running with version 1.79.1 - 47faaa4. Node version: 18.17.1
[1] prisma:error
[1] Invalid `prisma.user.delete()` invocation:
[1]
[1]
[1] Foreign key constraint failed on the field: `PetMedicalRecord_petId_fkey (index)`
[1] [2024-02-11T21:09:52.693] [ERROR] [SnailyCAD] - {
[1]   method: 'DELETE',
[1]   url: '/v1/admin/manage/users/cll7nxaom0001fb1eexkaqpb4',
[1]   route: '/v1/admin/manage/users/:id',
[1]   headers: {
[1]     connection: 'upgrade',
[1]     host: 'demo-api.snailycad.org',
[1]     'x-forwarded-for': '92.109.193.29',
[1]     'sec-ch-ua': '"Not_A Brand";v="8", "Chromium";v="120", "Opera GX";v="106"',
[1]     accept: 'application/json, text/plain, */*',
[1]     'sec-ch-ua-mobile': '?0',
[1]     'is-from-dispatch': 'false',
[1]     'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36 OPR/106.0.0.0',
[1]     session: '',
[1]     'sec-ch-ua-platform': '"Windows"',
[1]     origin: 'https://demo.snailycad.org',
[1]     'sec-fetch-site': 'same-site',
[1]     'sec-fetch-mode': 'cors',
[1]     'sec-fetch-dest': 'empty',
[1]     'accept-encoding': 'gzip, deflate, br',
[1]     'accept-language': 'nl-NL,nl;q=0.9,en-US;q=0.8,en;q=0.7',
[1]     cookie: 'google-analytics_v4_Snoo__ga4=44c3d74f-b4ce-4ff0-9c92-3c19b63c8b1f; google-analytics_v4_Snoo___z_ga_audiences=44c3d74f-b4ce-4ff0-9c92-3c19b63c8b1f; _ga=GA1.1.1805509624.1707076323; sn_isDarkTheme=true; google-analytics_v4_Snoo__session_counter=7; google-analytics_v4_Snoo__engagementDuration=0; _ga_HE45R7EFY2=GS1.1.1707609371.8.1.1707609812.53.0.0; google-analytics_v4_Snoo__engagementStart=1707609979591; google-analytics_v4_Snoo__counter=19; google-analytics_v4_Snoo__let=1707609979591; snaily-cad-session=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySWQiOiJjbHM3eHNhMTgwMDFuaTNrYzZ0czJzMXpoIiwiaWF0IjoxNzA3Njg1NzU0LCJleHAiOjE3MDc2ODkzNTR9.F9w3YXbon_V7YRSSC91mRNc89lUi8g8Ue7FKE-Ni3bw; snaily-cad-refresh-token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzZXNzaW9uSWQiOiJ2ZzhpOWl2YjRjZnZ2bmt2MW55azM2dzUiLCJpYXQiOjE3MDc2ODU3NTMsImV4cCI6MTcwODI5MDU1M30.ZaCFfZtCiK_SQnan249pH6THtquMdkW5xpKH8PZ3NWY'
[1]   },
[1]   body: {},
[1]   query: {},
[1]   params: {},
[1]   reqId: '96b7ec2a594a47bb90e880a3f4c801c5',
[1]   time: 2024-02-11T21:09:52.693Z,
[1]   duration: 1594,
[1]   error: {
[1]     name: 'PrismaClientKnownRequestError',
[1]     message: '\n' +
[1]       'Invalid `prisma.user.delete()` invocation:\n' +
[1]       '\n' +
[1]       '\n' +
[1]       'Foreign key constraint failed on the field: `PetMedicalRecord_petId_fkey (index)`',
[1]     status: 500,
[1]     errors: []
[1]   },
[1]   catch: true
[1] }
[1] [2024-02-11T21:09:52.698] [ERROR] [SnailyCAD] - {
[1]   method: 'DELETE',
[1]   url: '/v1/admin/manage/users/cll7nxaom0001fb1eexkaqpb4',
[1]   route: '/v1/admin/manage/users/:id',
[1]   headers: {
[1]     connection: 'upgrade',
[1]     host: 'demo-api.snailycad.org',
[1]     'x-forwarded-for': '92.109.193.29',
[1]     'sec-ch-ua': '"Not_A Brand";v="8", "Chromium";v="120", "Opera GX";v="106"',
[1]     accept: 'application/json, text/plain, */*',
[1]     'sec-ch-ua-mobile': '?0',
[1]     'is-from-dispatch': 'false',
[1]     'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36 OPR/106.0.0.0',
[1]     session: '',
[1]     'sec-ch-ua-platform': '"Windows"',
[1]     origin: 'https://demo.snailycad.org',
[1]     'sec-fetch-site': 'same-site',
[1]     'sec-fetch-mode': 'cors',
[1]     'sec-fetch-dest': 'empty',
[1]     'accept-encoding': 'gzip, deflate, br',
[1]     'accept-language': 'nl-NL,nl;q=0.9,en-US;q=0.8,en;q=0.7',
[1]     cookie: 'google-analytics_v4_Snoo__ga4=44c3d74f-b4ce-4ff0-9c92-3c19b63c8b1f; google-analytics_v4_Snoo___z_ga_audiences=44c3d74f-b4ce-4ff0-9c92-3c19b63c8b1f; _ga=GA1.1.1805509624.1707076323; sn_isDarkTheme=true; google-analytics_v4_Snoo__session_counter=7; google-analytics_v4_Snoo__engagementDuration=0; _ga_HE45R7EFY2=GS1.1.1707609371.8.1.1707609812.53.0.0; google-analytics_v4_Snoo__engagementStart=1707609979591; google-analytics_v4_Snoo__counter=19; google-analytics_v4_Snoo__let=1707609979591; snaily-cad-session=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySWQiOiJjbHM3eHNhMTgwMDFuaTNrYzZ0czJzMXpoIiwiaWF0IjoxNzA3Njg1NzU0LCJleHAiOjE3MDc2ODkzNTR9.F9w3YXbon_V7YRSSC91mRNc89lUi8g8Ue7FKE-Ni3bw; snaily-cad-refresh-token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzZXNzaW9uSWQiOiJ2ZzhpOWl2YjRjZnZ2bmt2MW55azM2dzUiLCJpYXQiOjE3MDc2ODU3NTMsImV4cCI6MTcwODI5MDU1M30.ZaCFfZtCiK_SQnan249pH6THtquMdkW5xpKH8PZ3NWY'
[1]   },
[1]   body: {},
[1]   query: {},
[1]   params: {},
[1]   reqId: '96b7ec2a594a47bb90e880a3f4c801c5',
[1]   time: 2024-02-11T21:09:52.698Z,
[1]   duration: 1599,
[1]   event: 'request.end',
[1]   status: 500,
[1]   status_code: '500',
[1]   state: 'KO',
[1]   error_name: 'PrismaClientKnownRequestError',
[1]   error_message: '\n' +
[1]     'Invalid `prisma.user.delete()` invocation:\n' +
[1]     '\n' +
[1]     '\n' +
[1]     'Foreign key constraint failed on the field: `PetMedicalRecord_petId_fkey (index)`',
[1]   error_stack: 'PrismaClientKnownRequestError: \n' +
[1]     'Invalid `prisma.user.delete()` invocation:\n' +
[1]     '\n' +
[1]     '\n' +
[1]     'Foreign key constraint failed on the field: `PetMedicalRecord_petId_fkey (index)`\n' +
[1]     '    at ni.handleRequestError (/root/snaily-cadv4/node_modules/.pnpm/@prisma+client@5.6.0_prisma@5.6.0/node_modules/@prisma/client/runtime/library.js:124:6817)\n' +
[1]     '    at ni.handleAndLogRequestError (/root/snaily-cadv4/node_modules/.pnpm/@prisma+client@5.6.0_prisma@5.6.0/node_modules/@prisma/client/runtime/library.js:124:6206)\n' +
[1]     '    at ni.request (/root/snaily-cadv4/node_modules/.pnpm/@prisma+client@5.6.0_prisma@5.6.0/node_modules/@prisma/client/runtime/library.js:124:5926)\n' +
[1]     '    at async l (/root/snaily-cadv4/node_modules/.pnpm/@prisma+client@5.6.0_prisma@5.6.0/node_modules/@prisma/client/runtime/library.js:129:10023)'
[1] }
`

--- a/apps/api/src/controllers/admin/manage/manage-users-controller.ts
+++ b/apps/api/src/controllers/admin/manage/manage-users-controller.ts
@@ -143,6 +143,24 @@ export class ManageUsersController {
     @BodyParams("userIds", String) userIds: string[],
     @BodyParams("days", Number) days = 30,
   ) {
+    await prisma.petMedicalRecord.deleteMany({
+      where: {
+        pet: {
+          citizen: {
+            userId: { in: userIds },
+          },
+        },
+      },
+    });
+
+    await prisma.pet.deleteMany({
+      where: {
+        citizen: {
+          userId: { in: userIds },
+        },
+      },
+    });
+
     const arr = await prisma.$transaction(
       userIds.map((id) =>
         prisma.user.deleteMany({
@@ -471,6 +489,24 @@ export class ManageUsersController {
     if (!user) {
       throw new NotFound("notFound");
     }
+
+    await prisma.petMedicalRecord.deleteMany({
+      where: {
+        pet: {
+          citizen: {
+            userId: user.id,
+          },
+        },
+      },
+    });
+
+    await prisma.pet.deleteMany({
+      where: {
+        citizen: {
+          userId: user.id,
+        },
+      },
+    });
 
     await prisma.user.delete({
       where: {

